### PR TITLE
Ruby and Asciidoctor support

### LINKED ISSUE #3: Support Asciidoctor
The issue I am currently working on.

Started from this [comment](https://github.com/gohugoio/hugo/pull/3799#issuecomment-322188801):

> I think it would be reasonable to extend the shimgo package to handle asciidoctor in a similar fashion.
>
>The cost of the old approach is in process creation, not in parsing so it'll work. If you want to add support in shimgo to handle a simple sinatra (or similar) ruby app with asciidoctor, it should work fine.

Depends on #2.

### PR DESCRIPTION
**Pull request by miltador**
Closes #3.

### CODE REVIEW

**Comment by miltador** [comment]
Rebased onto master. Ready for review.

**Comment by miltador** [comment]
@tychoish I didn't edit README, leave it to you cause it requires some restructuring and mentioning new formats and deps. I could then add new PR to just add info about new dependencies on `asciidoctor` and `sinatra`. I've put a `Gemfile` for easier setup of development environment so developer can do `bundle install` in project directory to install dependencies.

**Comment by miltador** [comment]
Did a benchmark with new shimmed approach and Hugo, generated 1,5k asciidoctor pages in 1 minute. Previously it took 2 minutes. Pretty good x2 speed boost!

I also uploaded the script source and a bit of documentation on it to [this repo](https://github.com/miltador/shimgo-ruby).

**Comment by miltador** [comment]
Played with it on a real data and tuned up configs to avoid errors. Consider it as tested :-)

**Comment by tychoish** [comment]
Looks good to me; I'll merge soon. 

It might be nice to have a script or something for updating shimgo from the shimgo-ruby repo or similar?

**Comment by miltador** [comment]
Yes, some kind of automation would be good. For me it is easier to maintain ruby service separately and then merge.

BTW added a possibility to specify extensions for Asciidoctor through env. variable to resolve gohugoio/hugo#3330.

File: files.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ import (
 )
 
 var pythonFiles map[string][]byte
+var rubyFiles map[string][]byte
```
Comment (tychoish) [inline files.go:14]: does it make sense to just have a "serviceFiles" map and have python and ruby files in one map rather than needing two structures?

**Review by tychoish — CHANGES_REQUESTED** [review]
Just have a minor quibble 

Also it seems like we could update hugo to have "asciidoc" and "asciidoctor" formats running independently, to avoid the need for the environment variable?

File: files.go (hunk 1 above)
Comment (miltador) [inline files.go]: Can be merged into one map. This was away from my eyes cause in IDE I just folded whole `pythonFiles` block to not to scroll it all and worked with `rubyFiles`.

**Comment by miltador** [comment]
> Also it seems like we could update hugo to have "asciidoc" and "asciidoctor" formats running independently, to avoid the need for the environment variable?

How will we achieve that? I mean differentiating one from another. `asciidoc` is a subset of `asciidoctor` cause last one supports first one. They do not differ by file extension too.

File: files.go (hunk 1 above)
Comment (miltador) [inline files.go]: Good idea. Done.

File: registry.go [hunk 2]
```diff
@@ -1,9 +1,13 @@
 package shimgo
 
 import (
+	"fmt"
+	"io/ioutil"
+	"os"
 	"sync"
 )
 
+var workingDirectory string
```
Comment (tychoish) [inline registry.go]: can we revert the change that requires having working directory be a package variable? I very much wanted to keep this encapsulated in the server implementation/internals. 

I think we can have one map of the service files without needing to have a global working directory (which makes this non-reentrant, in ways that I'd like to avoid.

File: registry.go [hunk 3]
```diff
@@ -30,6 +47,7 @@ func (s *servers) cleanup() {
 	for _, server := range s.backends {
 		server.stop()
 	}
+	os.RemoveAll(workingDirectory)
```
Comment (tychoish) [inline registry.go]: I think keep seperate directories, but have a single map with service file data internally.

**Review by tychoish — CHANGES_REQUESTED** [review]

**Comment by tychoish** [comment]
I like this, I'm willing to do the last of the integration work if you want me to.

**Comment by miltador** [comment]
Removed the last commit so you can do what you intent to and merge it. Thanks.

**Comment by tychoish** [comment]
Sorry for the delay, This is now merged and in master. 

Do you want to take over (either supplant or add a patch to mine?) https://github.com/gohugoio/hugo/pull/3799 to accommodate asciidoctor?

**Comment by miltador** [comment]
Ok, I will rebase and modify and then create a PR.
